File: server/src/routes/fuentes.ts
import { spawn } from 'node:child_process'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { Router } from 'express'
import { pool } from '../db.js'

export const fuentesRouter = Router()

const URL_LIMIT = 10 // mismo límite que muestra WebAnalyzerCard en el frontend

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const ANALYZE_SCRIPT = path.resolve(__dirname, '../../../scripts/analyze_news.py')
const PYTHON_BIN = process.env.PYTHON_BIN || 'python'

// Lanza scripts/analyze_news.py (fetch + readability + Claude) como
// subproceso, pasándole el lote por stdin. No se espera a que termine: el
// cliente hace polling vía GET /api/fuentes o GET /api/sugerencias.
function runAnalyzeNewsScript(fuentes: { id: string; url: string }[]): void {
  const proc = spawn(PYTHON_BIN, [ANALYZE_SCRIPT], { stdio: ['pipe', 'pipe', 'pipe'] })
  proc.stdin.write(JSON.stringify(fuentes))
  proc.stdin.end()
  proc.stdout.on('data', (chunk: Buffer) => console.log(`[analyze_news] ${chunk.toString().trim()}`))
  proc.stderr.on('data', (chunk: Buffer) => console.error(`[analyze_news] ${chunk.toString().trim()}`))
  proc.on('error', (err) => console.error('[analyze_news] no se pudo iniciar el proceso Python:', err))
}

fuentesRouter.post('/analizar', async (req, res) => {
  const urls: unknown = req.body?.urls
  if (!Array.isArray(urls) || urls.length === 0) {
    res.status(400).json({ error: 'Se requiere un arreglo "urls" con al menos un elemento' })
    return
  }
  if (urls.length > URL_LIMIT) {
    res.status(400).json({ error: `Máximo ${URL_LIMIT} URLs por lote` })
    return
  }

  const creados: { id: string; url: string }[] = []
  for (const rawUrl of urls) {
    if (typeof rawUrl !== 'string') continue

    let dominio: string
    try {
      dominio = new URL(rawUrl).hostname.replace(/^www\./, '')
    } catch {
      continue // URL inválida: se ignora silenciosamente
    }

    const { rows } = await pool.query<{ id: string }>(
      `INSERT INTO fuentes_noticias (url, dominio, estado) VALUES ($1, $2, 'procesando') RETURNING id`,
      [rawUrl, dominio],
    )
    creados.push({ id: rows[0].id, url: rawUrl })
  }

  if (creados.length > 0) {
    runAnalyzeNewsScript(creados)
  }

  res.status(202).json({ fuentes: creados })
})

fuentesRouter.get('/', async (_req, res) => {
  const { rows } = await pool.query(
    `SELECT id, url, dominio, estado, creado_en, procesado_en
     FROM fuentes_noticias ORDER BY creado_en DESC LIMIT 50`,
  )
  res.json(rows)
})

File: server/app/routes/fuentes.py
import json
import os
import subprocess
import threading
from pathlib import Path

from fastapi import APIRouter, Body, HTTPException
from starlette import status

from ..db import ejecutar_sp, get_cursor

router = APIRouter()

URL_LIMIT = 10  # mismo límite que muestra WebAnalyzerCard en el frontend

BASE_DIR = Path(__file__).resolve().parent.parent.parent.parent  # raíz del repo
ANALYZE_SCRIPT = BASE_DIR / "scripts" / "analyze_news.py"
PYTHON_BIN = os.getenv("PYTHON_BIN", "python")


def _run_analyze_news_script(fuentes: list[dict]) -> None:
    """Lanza scripts/analyze_news.py (fetch + readability + Claude) como
    subproceso, pasándole el lote por stdin. No se espera a que termine: el
    cliente hace polling vía GET /api/fuentes o GET /api/sugerencias."""

    def _run() -> None:
        try:
            proc = subprocess.Popen(
                [PYTHON_BIN, str(ANALYZE_SCRIPT)],
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
            )
            out, err = proc.communicate(input=json.dumps(fuentes))
            if out.strip():
                print(f"[analyze_news] {out.strip()}")
            if err.strip():
                print(f"[analyze_news] {err.strip()}")
        except Exception as exc:
            print(f"[analyze_news] no se pudo iniciar el proceso Python: {exc}")

    threading.Thread(target=_run, daemon=True).start()


@router.post("/analizar", status_code=status.HTTP_202_ACCEPTED)
def analizar(body: dict = Body(...)):
    urls = body.get("urls")
    if not isinstance(urls, list) or len(urls) == 0:
        raise HTTPException(status_code=400, detail={"error": 'Se requiere un arreglo "urls" con al menos un elemento'})
    if len(urls) > URL_LIMIT:
        raise HTTPException(status_code=400, detail={"error": f"Máximo {URL_LIMIT} URLs por lote"})

    # "creados": URLs nuevas (o un reintento de una que había fallado antes)
    # que sí se van a analizar. "omitidas": URLs que ya se habían enviado y
    # siguen en proceso o ya se analizaron con éxito -- no se vuelven a
    # analizar ni se les crea una fila duplicada (ver sp_fuentes_crear).
    creados: list[dict] = []
    omitidas: list[dict] = []
    with get_cursor() as cur:
        for raw_url in urls:
            if not isinstance(raw_url, str):
                continue

            try:
                from urllib.parse import urlparse

                parsed = urlparse(raw_url)
                if not parsed.scheme or not parsed.netloc:
                    continue
                dominio = parsed.hostname or ""
                if dominio.startswith("www."):
                    dominio = dominio[4:]
            except Exception:
                continue  # URL inválida: se ignora silenciosamente

            ejecutar_sp(cur, "SELECT * FROM sp_fuentes_crear(%s, %s)", (raw_url, dominio))
            fila = cur.fetchone()
            if fila["debe_analizar"]:
                creados.append({"id": fila["id"], "url": raw_url})
            else:
                omitidas.append({"url": raw_url, "estado": fila["estado_actual"]})

    if creados:
        _run_analyze_news_script(creados)

    return {"fuentes": creados, "omitidas": omitidas}


@router.get("")
def listar():
    with get_cursor() as cur:
        ejecutar_sp(cur, "SELECT * FROM sp_fuentes_listar()")
        return cur.fetchall()
